### ISSUE TITLE
Improvement notification export for GDExtensions

### ISSUE THREAD

**Issue by snatvb** [godotengine/godot#93366]
### Tested versions

- 4.2 stable

### System information

Windows 11 - 4.2

### Issue description

If user tries convert `@export_file("*.txt")` to Array:
```gdscript
extends Node

class_name NewScript

@export_file("*.txt") var files: Array[String]
```
LSP shows error:
![image](https://github.com/godot-rust/gdext/assets/1572963/4dabb2ac-4877-4901-b003-c9ff6892bb25)
And in console:
![image](https://github.com/godot-rust/gdext/assets/1572963/5646095d-a992-49d2-ab98-738840f9773b)

But if user does it in rust:
```rs
#[derive(GodotClass)]
#[class(init, base=Node)]
struct Foo {
    base: Base<Node>,
    #[export(file)]
    resources: Array<GString>,
}

#[godot_api]
impl INode for Foo {}
```

It leads to this:
![image](https://github.com/godot-rust/gdext/assets/1572963/a9172443-b756-4d4a-a7a7-d37734624582)
![image](https://github.com/godot-rust/gdext/assets/1572963/2db4c31a-6431-4b46-afbf-09945905f3cd)

For user it's unexpected and it would be rather to notify about incorrect annotation.

### Steps to reproduce

- Add extenstion
- Describe incorrect export

### Minimal reproduction project (MRP)

[godot-rust-repro.zip](https://github.com/user-attachments/files/15904998/godot-rust-repro.zip)

**Comment by lilizoey** [comment]
i dont think this is a godot issue? isn't this just an issue with the rust bindings? im not sure what godot could do here.

**Comment by Bromeon** [comment]
As I understood, there are two things:

1. A feature request/discussion whether `@export_file("*.filter")` could potentially be allowed on `Array` -- visualized as a list in UI.
2. If not: the question whether the error message _"@export_file" requires a variable of type "String" but type "XY" was given instead_ is exclusive to the GDScript parser, or it can/could also be used for GDExtension properties.

On Rust side, we should in parallel make sure that the property is registered properly with Godot.

**Comment by dsnopek** [comment]
> 2\. If not: the question whether the error message _"@export_file" requires a variable of type "String" but type "XY" was given instead_ is exclusive to the GDScript parser, or it can/could also be used for GDExtension properties.

Currently, this message comes from `gdscript_parser.cpp`.

I suppose we could put similar validation in `ClassDB::add_property()`, however, the error message couldn't really be as good.

From Godot's perspective, it's looking at data passed on the `PropertyInfo` struct, and the error message could only refer to that (for example, saying the `PropertyInfo::hint` won't work with the given `PropertyInfo::type`). However, the binding probably doesn't directly expose `PropertyInfo`, but instead some binding-specific way of doing it, and so it'd be able to provide a better error message that refers to that.

So, I personally think giving a nice error message should be up to the GDExtension binding (in this case, the Rust binding).

**Comment by dsnopek** [comment]
Discussed at the GDExtension meeting:

- It would be nice to have an API that that could check if a property is valid, that could potentially be used by GDScript, `ClassDB` and GDExtension bindings, and still provide nice error messages.
- However, some bindings may be able to check for property validity at compile time (making this unnecessary)
- The property hints have [decent documentation for](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#enum-globalscope-propertyhint) but we'd need to make sure it's complete and stays that way.
- We'd also need to ensure that the potential future validation API would remain in sync.
- For now, we recommend that GDExtension bindings attempt to handle this validation themselves, based on the documentation linked above.

**Comment by dsnopek** [comment]
> could we get it for native bindings?

It is possible for native bindings, so long as the language binding exposes it somehow. In godot-cpp, this would be like:

```
ADD_PROPERTY(PropertyInfo(Variant::STRING, "file", PROPERTY_HINT_FILE, "*.txt"), "set_file", "get_file");
```

**Comment by snatvb** [comment]
As I get it works fine now, @Bromeon added support. I'll close the issue because it's solved. Thanks dears @dsnopek & @Bromeon!
